README.md
<div align="center" markdown="1">

# [`app-cli-autoscroll`][url-repo]

[![License][badge-license]][url-license]
[![Status][badge-status-finished]][url-repo]
[![Version][badge-version]][url-version]
[![Release][badge-workflow-release]][url-workflow-release]
[![Conventional Commits][badge-conventionalcommits]][url-conventionalcommits]

Cross-platform CLI app enabling universal autoscroll

</div>

## About The Project

I bought a Logitech Marble Mouse and couldn't make the autoscroll work on Linux, 
so I decided to make a python package to solve this problem 

It was a pretty much pointless idea because on Linux you can achieve autoscroll using config files
(see [the example](#xorg-server-config-example)) and for Windows there are usually drivers,
but I decided to finish the project anyway

The package supports only mouse buttons

The package requires `pynput` to work, to display an icon you need to install `pyside6` (`qt6` package)

```bash
pip install autoscroll pyside6
```

Built-in icon: <img src="https://raw.githubusercontent.com/shishifubing-com/app-cli-autoscroll/master/autoscroll/resources/img/icon.svg" width="50" height="50">

You can run it on startup by creating a systemd service on Linux or by adding
the `.exe` file of the script in the venv folder to autostart on Windows

## Usage

By default, the icon is disabled, to enable it pass `--icon-enable`

You can pass file contents as command line arguments using `@path/to/the/file` syntax.
Arguments in that case can be placed wherever - on one line, on several lines

If you want to dynamically pass runtime arguments (without restarting the process), you can use `--config` options for it

Once you press `--buttons-start`, you can scroll vertically or horizontally just by moving your mouse untill you press `--buttons-end`

If `--buttons-hold` is set, the srolling ends once you release `--buttons-start`

Once `--buttons-start` is pressed, the scroll thread starts looping
Every loop consists of sleeping for an interval, then scrolling for either 0, 1, or -1 pixels on both axis towards the starting point
Starting point is the point where `--buttons-start` was pressed
Sleep interval is recalculated on every mouse move as such:

```
    100 / (--scrolling-acceleration * max(distance) + --scrolling-speed)
```

If `--scrolling-acceleration` is not 0, the speed of scrolling will be faster
the farther away you are from the starting point
If `--scrolling-acceleration` is 0, the speed of scrolling will be constant

### Examples

#### Use the package

```bash
python3 -m venv venv
. venv/bin/activate
pip install autoscroll pyside6
autoscroll
```

#### Use the source code

```bash
python3 -m venv venv
. venv/bin/activate
git clone https://github.com/shishifubing/app-cli-autoscroll.git
pip install -r autoscroll/requirements.txt
python3 -m autoscroll
```

#### Command line options

```bash
autoscroll --buttons-start 1 --debug-click --icon-disable
```

#### Pass a configuration file once

```bash
autoscroll --icon-enable @config.txt
```

If `config.txt` is defined like this, its contents will be used as command line arguments - they will be loaded only once
Arguments can be placed wherever - on one line, on several lines
For example,

```
--buttons-start 1
--buttons-hold --debug-click
```

#### Listen for changes in the configuration file

```bash
autoscroll --config-enable --config-path config.txt
```

If config.txt is defined like this, the process will listen for changes in that
file and update itself
Arguments can be placed wherever - on one line, on several lines
The file is checked for changess every `--config-interval`
For example:

```
--buttons-start 1 --buttons-hold
--debug_click
```

### `--help` output

```
usage: autoscroll [-h] [-ss SCROLLING_SPEED] [-sd SCROLLING_DEAD_AREA]
                  [-sa SCROLLING_ACCELERATION] [-bh] [-bs BUTTONS_START] [-be BUTTONS_END]
                  [-ce] [-cp CONFIG_PATH] [-ci CONFIG_INTERVAL] [-ie] [-ip ICON_PATH]
                  [-is ICON_SIZE] [-df] [-dc] [-ds] [-di]

...

options:
  -h, --help            show this help message and exit

scrolling:

  -ss, --scrolling-speed int
                        constant part of the scrolling speed
                        [default: 300]
  -sd, --scrolling-dead-area int
                        size of the square area aroung the starting point where scrolling will stop, in
                        pixels
                        [default: 50]
  -sa, --scrolling-acceleration int
                        dynamic part of the scrolling speed, depends on the distance from the point
                        where the scrolling started, can be set to 0
                        [default: 10]

buttons:

  -bh, --buttons-hold   if set, the scrolling will end once you release --buttons-start
  -bs, --buttons-start int
                        button that starts the scrolling
                        [default: 2]
  -be, --buttons-end int
                        button that ends the scrolling
                        [default: --buttons-start]

config:

  -ce, --config-enable  if set, arguments from the configuration file on --config-path will be loaded
                        every --config-interval
  -cp, --config-path str
                        path to the configuration file
                        [default: /home/kongrentian/.config/autoscroll/config.txt]
  -ci, --config-interval int
                        how often the config file should be checked for changes, in seconds
                        [default: 5]

icon:

  -ie, --icon-enable    if set, the icon will be enabled
  -ip, --icon-path str  path to the icon
                        [default: resources/img/icon.svg]
  -is, --icon-size int  size of the icon, in pixels
                        [default: 30]

debug:

  -df, --debug-file     if set, every time the config file is parsed, information will be printed to
                        stdout
  -dc, --debug-click    if set, click info will be printed to stdout
  -ds, --debug-scroll   if set, scroll info will be printed to stdout
  -di, --debug-initial  if set, startup configuration will be printed to stdout
```

## xorg-server config example

```conf
# https://wiki.archlinux.org/title/Logitech_Marble_Mouse#Configuration_file
# https://help.ubuntu.com/community/Logitech_Marblemouse_USB
# https://bbs.archlinux.org/viewtopic.php?id=261138
# path - /etc/X11/xorg.conf.d/40-trackball.conf
Section "InputClass"
  Identifier   "Marble Mouse"
  MatchProduct "Logitech USB Trackball"
  Driver       "libinput"
  Option       "ScrollMethod"     "button"
  Option       "ScrollButton"     "1"
  Option       "MiddleEmulation"  "true"
  Option       "ButtonMapping"    "3 2 1 4 5 6 7 9 8"
  Option       "NaturalScrolling" "true"
EndSection
```


<!-- relative links -->

<!-- project links -->

[url-repo]: https://github.com/shishifubing-com/app-cli-autoscroll
[url-license]: https://github.com/shishifubing/app-cli-autoscroll/blob/main/LICENSE
[url-workflow-release]: https://github.com/shishifubing/app-cli-autoscroll/actions/workflows/release.yml
[url-version]: https://github.com/shishifubing/app-cli-autoscroll/releases/latest


<!-- external links -->

[url-conventionalcommits]: https://conventionalcommits.org

<!-- badge links -->

[badge-status-finished]: https://img.shields.io/badge/status-finished-informational
[badge-license]: https://img.shields.io/github/license/shishifubing/app-cli-autoscroll.svg
[badge-workflow-release]: https://img.shields.io/github/actions/workflow/status/shishifubing/app-cli-autoscroll/release.yml?branch=main&label=release&logo=github
[badge-version]: https://img.shields.io/github/v/release/shishifubing/app-cli-autoscroll?label=version

<!-- other badge links -->

[badge-conventionalcommits]: https://img.shields.io/badge/conventional--commits-1.0.0-%23FE5196?logo=conventionalcommits&logoColor=white


CHANGELOG.md
# Changelog

## v0.1.35 - 2023-03-14

<!-- Automatically generated in https://github.com/shishifubing/app-cli-autoscroll/actions/runs/4421081136 -->
**Full Changelog**: https://github.com/shishifubing/app-cli-autoscroll/compare/v0.1.34...v0.1.35

## v0.1.34 - 2023-03-13

### Documentation changes

- **readme**: center the header by @tiandaoburen in [#2](https://github.com/shishifubing/app-cli-autoscroll/pull/2)
- **readme**: change headings by @tiandaoburen in [#3](https://github.com/shishifubing/app-cli-autoscroll/pull/3)

**Full Changelog**: https://github.com/shishifubing/app-cli-autoscroll/compare/v0.1.0...v0.1.34


autoscroll/autoscroll/constants.py
from typing import Any, Dict
from os import environ as os_environ
from .functions import get_resource_content
from .arguments import ArgparseFormatter

SCROLLING_SPEED: int = 300
SCROLLING_ACCELERATION_DISTANCE: int = 10
SCROLLING_SLEEP_INTERVAL_INITIAL: float = 0.1
SCROLLING_DEAD_AREA: int = 50

BUTTONS_START: int = 2
BUTTONS_HOLD: bool = False

ICON_ENABLE: bool = False
ICON_SIZE: int = 30
ICON_PATH: str = 'resources/img/icon.svg'
ICON_ERROR: str = ('icon is enabled (it is disabled by default), but the '
                   '\'pyside6\' package is not installed. '
                   'remove \'--icon-enable\' or install the package\n'
                   '\npip install pyside6\n')

CONFIG_PATH: str = f'{os_environ.get("HOME")}/.config/autoscroll/config.txt'
CONFIG_ENABLE: bool = False
CONFIG_INTERVAL: int = 5
CONFIG_ERROR_ENABLE: str = 'you are trying to enable the config (\'enable\' is set to \'True\'), but the path is not valid'
CONFIG_ERROR_PARSE: str = 'you are trying to parse the config file, but \'enable\' is \'False\''


DEBUG_SCROLL: bool = False
DEBUG_CLICK: bool = False
DEBUG_INITIAL: bool = False
DEBUG_FILE: bool = False
DEBUG_PADDING: int = 16

COORDINATE_NAME: str = 'coordinates'


PARSER_INITIALIZER: Dict[str, Any] = {
    'prog': 'linux-xorg-autoscroll',
    'formatter_class': ArgparseFormatter,
    'description': get_resource_content('resources/txt/prolog.txt'),
    'fromfile_prefix_chars': '@'}

ARGUMENTS: Dict[str, Any] = {
    'scrolling': {
        'speed': {
            'type': int,
            'help': ('R|constant part of the scrolling speed\n'
                     f'[default: {SCROLLING_SPEED}]'),
        },
        'dead-area': {
            'type': int,
            'help': ('R|size of the square area aroung the starting '
                     'point where scrolling will stop, in pixels\n'
                     f'[default: {SCROLLING_DEAD_AREA}]')
        },
        'acceleration': {
            'type': int,
            'help': ('R|dynamic part of the  scrolling speed, depends on the '
                     'distance from the point where the scrolling started, '
                     'can be set to 0\n'
                     f'[default: {SCROLLING_ACCELERATION_DISTANCE}]')
        }
    },
    'buttons': {
        'hold': {
            'action': 'store_const',
            'const': True,
            'help': ('R|if set, the scrolling will end once you release '
                     '--buttons-start')
        },
        'start': {
            'type': int,
            'help': f'R|button that starts the scrolling\n[default: {BUTTONS_START}]'
        },
        'end': {
            'type': int,
            'help': ('R|button that ends the scrolling\n'
                     '[default: --buttons-start]')
        }
    },
    'config': {
        'enable': {
            'action': 'store_const',
            'const': True,
            'help': ('R|if set, arguments from the configuration file '
                     'on --config-path will be loaded every --config-interval')
        },
        'path': {
            'type': str,
            'help': f'R|path to the configuration file\n[default: {CONFIG_PATH}]',
        },
        'interval': {
            'type': int,
            'help': ('R|how often the config file should be checked for '
                     f'changes, in seconds\n[default: {CONFIG_INTERVAL}]')
        }
    },
    'icon': {
        'enable': {
            'action': 'store_const',
            'const': True,
            'help': 'if set, the icon will be enabled'
        },
        'path': {
            'type': str,
            'help': f'R|path to the icon\n[default: {ICON_PATH}]'
        },
        'size': {
            'type': int,
            'help': f'R|size of the icon, in pixels\n[default: {ICON_SIZE}]'
        }
    },
    'debug': {
        'file': {
            'action': 'store_const',
            'const': True,
            'help': ('if set, every time the config file is parsed, '
                     'information will be printed to stdout')
        },
        'click': {
            'action': 'store_const',
            'const': True,
            'help': 'if set, click info will be printed to stdout'
        },
        'scroll': {
            'action': 'store_const',
            'const': True,
            'help': 'if set, scroll info will be printed to stdout'
        },
        'initial': {
            'action': 'store_const',
            'const': True,
            'help': 'if set, startup configuration will be printed to stdout'
        }
    }
}
